crates/matrix-sdk-crypto/src/session_manager/sessions.rs

```diff
@@ -235,16 +235,19 @@ impl SessionManager {
         )
         .await?;
 
+        // A map from user ID to a map from device ID to list of algorithms.
+        let mut non_olm_devices_by_user: BTreeMap<_, BTreeMap<_, _>> = BTreeMap::new();
+
+        // A map from user id to a set of device IDs.
+        let mut bad_key_devices: BTreeMap<_, BTreeSet<_>> = BTreeMap::new();
+
         for (user_id, user_devices) in devices_by_user {
             for (device_id, device) in user_devices {
                 if !(device.supports_olm()) {
-                    warn!(
-                        user_id = ?device.user_id(),
-                        device_id = ?device.device_id(),
-                        algorithms = ?device.algorithms(),
-                        "Device doesn't support any of our 1-to-1 E2EE \
-                        algorithms, can't establish an Olm session"
-                    );
+                    non_olm_devices_by_user
+                        .entry(user_id.clone())
+                        .or_default()
+                        .insert(device_id, Vec::from(device.algorithms()));
```

Comment:
rather than logging each device separately, we collect a map of the affected devices and log them at the end.